Catching polymorphic type 'struct gsl::fail_fast' by value

When building with GCC 8.3.0 and using `-Wcatch-value -Werror` I get the following errors:

```
In file included from /build/source/tests/multi_span_tests.cpp:24:
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____4()':
/build/source/tests/multi_span_tests.cpp:155:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:158:51: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(const_workaround_macro(), fail_fast);
                                                   ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:163:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:166:51: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(const_workaround_macro(), fail_fast);
                                                   ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____8()':
/build/source/tests/multi_span_tests.cpp:262:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____10()':
/build/source/tests/multi_span_tests.cpp:295:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:301:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:307:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:313:45: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(workaround_macro(), fail_fast);
                                             ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____14()':
/build/source/tests/multi_span_tests.cpp:455:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s[10][3][4], fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____28()':
/build/source/tests/multi_span_tests.cpp:772:47: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.first(6).length(), fail_fast);
                                               ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____30()':
/build/source/tests/multi_span_tests.cpp:815:46: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.last(6).length(), fail_fast);
                                              ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____32()':
/build/source/tests/multi_span_tests.cpp:851:52: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(0, 6).length(), fail_fast);
                                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:852:52: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(1, 5).length(), fail_fast);
                                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:860:52: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(6, 0).length(), fail_fast);
                                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:868:56: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS((av.subspan<1, 0>().length()), fail_fast);
                                                        ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:874:49: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(1).length(), fail_fast);
                                                 ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:883:49: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(6).length(), fail_fast);
                                                 ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:894:49: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(av.subspan(6).length(), fail_fast);
                                                 ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____36()':
/build/source/tests/multi_span_tests.cpp:929:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s.extent(1), fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:939:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s.extent(1), fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:951:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s.extent(3), fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:963:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s.extent(3), fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____38()':
/build/source/tests/multi_span_tests.cpp:974:31: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(s(5), fail_fast);
                               ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____44()':
/build/source/tests/multi_span_tests.cpp:1133:35: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(av[1][3] = 3, fail_fast);
                                   ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1134:39: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS((av[{1, 3}] = 3), fail_fast);
                                       ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1136:32: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(av[10][2], fail_fast);
                                ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1137:36: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS((av[{10, 2}]), fail_fast);
                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1139:32: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(av[-1][0], fail_fast);
                                ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1140:36: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS((av[{-1, 0}]), fail_fast);
                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1142:32: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(av[0][-1], fail_fast);
                                ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1143:36: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS((av[{0, -1}]), fail_fast);
                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____52()':
/build/source/tests/multi_span_tests.cpp:1295:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av[0], fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1296:46: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av.begin()[0], fail_fast);
                                              ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1297:47: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av.cbegin()[0], fail_fast);
                                               ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1308:38: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av[0], fail_fast);
                                      ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1309:46: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av.begin()[0], fail_fast);
                                              ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1310:47: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(empty_av.cbegin()[0], fail_fast);
                                               ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____64()':
/build/source/tests/multi_span_tests.cpp:1616:36: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(av1[10][3][4], fail_fast);
                                    ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____66()':
/build/source/tests/multi_span_tests.cpp:1710:30: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
         CHECK_THROWS_AS(f(), fail_fast);
                              ^~~~~~~~~
/build/source/tests/multi_span_tests.cpp:1719:26: error: catching polymorphic type 'struct gsl::fail_fast' by value [-Werror=catch-value=]
     CHECK_THROWS_AS(f(), fail_fast);
                          ^~~~~~~~~
```

--- a/pkgs/development/libraries/microsoft_gsl/default.nix
+++ b/pkgs/development/libraries/microsoft_gsl/default.nix
@@ -20,6 +20,9 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ catch cmake ];
   buildPhase = if nativeBuild then "make" else "true";
 
+  # https://github.com/microsoft/GSL/issues/806
+  cmakeFlags = [ "-DCMAKE_CXX_FLAGS=-Wno-catch-value" ];
+
   installPhase = ''
     mkdir -p $out/include
     mv ../include/ $out/

